build_tools/pkgci/bisect/README.md

```diff
@@ -0,0 +1,329 @@
+packages (i.e. `iree-base-compiler` and `iree-base-runtime`) and prepend their
+installed location to the `PATH` environment variable. If you want to run a
+bisect that _does not_ need to run `iree-compile`, just use `git bisect`
+directly. However, if you _do_ need to run `iree-compile`, this can save
+substantial time by avoiding the need to build it from source at each test
+commit.
+
+## Prerequisites
+
+### System requirements
+
+Requirement | Details
+----------- | -------
+Linux | (at least until IREE builds packages for other systems at each commit)
+`git` | https://git-scm.com/
+`gh` CLI | https://cli.github.com/
+iree-org/iree repository read access | Needed to [download workflow artifacts](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/downloading-workflow-artifacts). See also [obtaining commit access](https://iree.dev/developers/general/contributing/#obtaining-commit-access).
+`python3.11` with `venv` support | (Version must match what PkgCI builds) `sudo apt install python3.11 python3.11-dev python3.11-venv`
+
+### Data requirements
+
+* A command to run, such as a `.mlir` file and a `iree-compile` command
+* A known-working commit
+* A known-broken commit
+
+The commit range between known-working and known-broken should be as small as
+possible to limit the number of test steps. The bisect algorithm is `O(log N)`
+where `N` is the number of commits in the range, but wider windows have larger
+risk of something in the test environment breaking (e.g. breaking API changes,
+serialized `.mlir` files not being stable, etc.).
+
+## Usage
+
+### Example
+
+Let's try to find the culprit commit for issue
+https://github.com/iree-org/iree/issues/18879. Thanks to the detailed issue
+description, we have all the data we need to run a bisect already.
+
+To run the bisect tool:
+
+1. Setup the test case by saving the input `.mlir` file and constructing a test
+   `.sh` script:
+
+    ```mlir
+    // /tmp/issue_18879.mlir
+
+    // This is the input program to the test script.
+    module {
+      func.func @main_graph(
+        // ...
+    ```
+
+    ```bash
+    # /tmp/issue_18879.sh
+
+    # This is the script that will be tested at each commit.
+    # This script should succeed (return 0) at and prior to the `--good-ref`
+    # commit and should fail (return non-0) at the `--bad-ref` commit.
+
+    # Try to keep these test scripts minimal. If the failure is in an earlier
+    # phase of the compiler (e.g. 'Flow' or 'Stream'), consider using
+    # a flag like `--compile-to=hal` to exit early on successful run instead
+    # of spending all the time to serialize an output `.vmfb` file.
+    # https://iree.dev/developers/general/developer-tips/#compiling-phase-by-phase
+
+    iree-compile --iree-hal-target-backends=llvm-cpu -o /dev/null /tmp/issue_18879.mlir
+    ```
+
+    ```bash
+    # Set the script as executable.
+    chmod +x /tmp/issue_18879.sh
+    ```
```

Comment:
(for a future PR...)

For the simple case where the test script is a single command, we can elide this step via a new `--test-command=` parameter, like so:

```bash
    ./bisect_packages.py \
      --good-ref=f9fa934c649749b30fc4be05d9cef78eb043f0e9 \
      --bad-ref=05bbcf1385146d075829cd940a52bf06961614d0 \
      --test-command="iree-compile --iree-hal-target-backends=llvm-cpu -o /dev/null /tmp/issue_18879.mlir"
```